### PR TITLE
Feature/mail publisher

### LINKED ISSUE #103: Mail-Publisher
Erstellte Reports über ein Product automatisiert über eine Email-Vorlage (Adressat, Betreff, plain-text oder pdf als attachment, Formatierung) versenden
___________________________

Send created reports about a product  - automatically via an email template (recipient, subject, plaintext or pdf as attachment, formatting).

Reqs:
- No encryption for now (PGP, etc.)
- Send an email to a preconfigured email address
- Add auto email option to settings to preconfigure the email address, and/or to enable/disable automatic email sending.
- It would most probably appear on the publisher level (email address and possible signing using PGP)

Qs:   
- How to trigger the sending - once in a time like every 30 minutes it would check new products, so the new products are not sent with errors.

Consider: https://github.com/CZ-NIC/envelope/issues/29#issue-1780690838

### PR DESCRIPTION
**Pull request by not4Pedro**
resolves #103 

## Summary by CodeRabbit

- **New Features**
    - Introduced configurations for email publishers and product rendering data for vulnerability reports.
    - Added test cases for an email publisher functionality, including publishing emails with and without SMTP server addresses and user inputs.

### CODE REVIEW

**Comment by not4Pedro** [comment]
@coderabbitai review

File: src/worker/worker/publishers/email_publisher.py [hunk 1]
```diff
@@ -1,136 +1,134 @@
 import datetime
 import smtplib
-from email.message import Message
-from email.mime.base import MIMEBase
-from email.mime.multipart import MIMEMultipart
-from email.mime.text import MIMEText
-import gnupg
+import logging
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:3]: Please use the centralized logger from worker.log

File: src/worker/worker/publishers/email_publisher.py [hunk 2]
```diff
@@ -1,136 +1,134 @@
...
+
+        file_name = f"product_{datetime.datetime.now().strftime('%d-%m-%Y_%H-%M')}"
+        maintype, subtype = rendered_product.mime_type.split("/")
+        logger.info("EMAIL Publisher: Creating attachment")
+        attachment_data = base64.b64decode(rendered_product.data)
+        self.msg.add_attachment(attachment_data, maintype=maintype, subtype=subtype, filename=f"{file_name}")
+
+    def send_email(self, rendered_product: Product):
+        if rendered_product.mime_type in ["text/plain", "text/html"]:
+            # if rendered_product.mime_type in [""]:
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:95]: seems obsolete

File: src/worker/worker/publishers/email_publisher.py [hunk 3]
```diff
@@ -1,136 +1,134 @@
...
+            return {"error": "No SMTP server port provided"}
+
+        if smtp_server_tls := publisher.get("parameters").get("SERVER_TLS", None):
+            self.smtp_server_tls = smtp_server_tls
+
+        if smtp_server_username := publisher.get("parameters").get("EMAIL_USERNAME"):
+            self.smtp_server_username = smtp_server_username
+
+        if smtp_server_password := publisher.get("parameters").get("EMAIL_PASSWORD"):
+            self.smtp_server_password = smtp_server_password
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:48]: It seems like the credentials aren't used

File: src/worker/worker/publishers/email_publisher.py [hunk 4]
```diff
@@ -1,136 +1,134 @@
...
+        self.smtp_server_recipient = None
+        self.smtp_server_subject = None
+
+        self.msg = None
+        self.type = "EMAIL_PUBLISHER"
+        self.name = "EMAIL Publisher"
+        self.description = "Publisher for publishing by email"
+
+    def publish(self, publisher, publisher_input) -> dict[str, str] | None:
+        print(publisher)
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:30]: remove

File: src/worker/worker/publishers/email_publisher.py [hunk 5]
```diff
@@ -1,136 +1,134 @@
...
+        self.smtp_server_sender = None
+        self.smtp_server_recipient = None
+        self.smtp_server_subject = None
+
+        self.msg = None
+        self.type = "EMAIL_PUBLISHER"
+        self.name = "EMAIL Publisher"
+        self.description = "Publisher for publishing by email"
+
+    def publish(self, publisher, publisher_input) -> dict[str, str] | None:
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:29]: Please add a simple unit test for this method

**Review by b3n4kh — CHANGES_REQUESTED** [review]

File: src/worker/worker/publishers/email_publisher.py [hunk 6]
```diff
@@ -1,136 +1,160 @@
...
+            logger.error(f"Your SMTP server throws an error: {str(e)}")
+            return "An SMTP error occurred"
+
+    def send_with_tls(self, context):
+        try:
+            with SMTP_SSL(self.smtp_address, self.smtp_port, context=context) as server:
+                return self.send_mail(server)
+        except Exception as e:
+            logger.error(f"Your SMTP server with TLS context throws an error: {str(e)}")
+            return "Your SMTP server throws an error"
```
[the remark this answers is not in the document]
Comment (not4Pedro) [reply src/worker/worker/publishers/email_publisher.py:109]: IMO this is not feasible

File: src/worker/worker/publishers/email_publisher.py [hunk 7]
```diff
@@ -1,136 +1,160 @@
...
 
+    def send_without_tls(self):
+        try:
+            server = smtplib.SMTP(self.smtp_address, self.smtp_port)
+            return self.send_mail(server)
+        except Exception as e:
+            logger.error(f"Your SMTP server throws an error: {str(e)}")
+            return "Your SMTP server throws an error"
```
[the remark this answers is not in the document]
Comment (not4Pedro) [reply src/worker/worker/publishers/email_publisher.py:117]: IMO this is not feasible

File: src/worker/worker/publishers/email_publisher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (not4Pedro) [reply src/worker/worker/publishers/email_publisher.py:109]: For TLS connection a context is created and for non TLS no context is created

File: src/worker/worker/publishers/email_publisher.py [hunk 8]
```diff
@@ -1,136 +1,156 @@
...
-
-        except Exception as error:
-            BasePublisher.print_exception(self, error)
+        return {"message": "Email Publisher: Task Successful"}
+
+
+if __name__ == "__main__":
+    email_publisher = EMAILPublisher()
+    email_publisher.publish(
+        # {
```
Comment (b3n4kh) [inline src/worker/worker/publishers/email_publisher.py:132]: Move this code (as it is?) to the tests folder

File: src/core/core/managers/pre_seed_data.py [hunk 9]
```diff
@@ -201,14 +201,14 @@
...
-            "EMAIL_SUBJECT",
-            "EMAIL_MESSAGE",
-            "EMAIL_ENCRYPTION",
+            {"parameter": "SMTP_SERVER_ADDRESS", "rules": "required"},
+            {"parameter": "SMTP_SERVER_PORT"},
+            {"parameter": "SERVER_TLS", "type": "switch"},
+            {"parameter": "EMAIL_USERNAME"},
+            {"parameter": "EMAIL_PASSWORD"},
+            {"parameter": "EMAIL_SENDER", "rules": "email"},
+            {"parameter": "EMAIL_RECIPIENT", "rules": "email"},
```
Comment (b3n4kh) [inline src/core/core/managers/pre_seed_data.py:210]: I think sender & recipient should be required fields.

File: src/core/core/managers/pre_seed_data.py [hunk 10]
```diff
@@ -201,14 +201,14 @@
...
-            "SMTP_SERVER",
-            "SMTP_SERVER_PORT",
-            "EMAIL_USERNAME",
-            "EMAIL_PASSWORD",
-            "EMAIL_RECIPIENT",
-            "EMAIL_SUBJECT",
-            "EMAIL_MESSAGE",
-            "EMAIL_ENCRYPTION",
+            {"parameter": "SMTP_SERVER_ADDRESS", "rules": "required"},
+            {"parameter": "SMTP_SERVER_PORT"},
```
Comment (b3n4kh) [inline src/core/core/managers/pre_seed_data.py:205]: port => type: number

**Review by b3n4kh — CHANGES_REQUESTED** [review]

File: src/core/core/managers/pre_seed_data.py (hunk 9 above)
Comment (not4Pedro) [reply src/core/core/managers/pre_seed_data.py:210]: ideally it would be both

**Review by b3n4kh — APPROVED** [review]

**Pull request merged**